sdk/core/core/src/az_log.c

```diff
@@ -10,62 +10,55 @@
 #include <az_log_internal.h>
 #include <az_span.h>
 
-#include <stdbool.h>
 #include <stddef.h>
 
 #include <_az_cfg.h>
 
 static az_log_classification const* _az_log_classifications = NULL;
-static size_t _az_log_classifications_length = 0;
-static az_log_fn _az_log_listener = NULL;
+static az_log_message_fn _az_log_message_callback = NULL;
 
-void az_log_set_classifications(
-    az_log_classification const* classifications,
-    size_t classifications_length)
+void az_log_set_classifications(az_log_classification const classifications[])
 {
-  // TODO: thread safety
   _az_log_classifications = classifications;
-  _az_log_classifications_length = classifications_length;
 }
 
-void az_log_set_listener(az_log_fn listener)
+void az_log_set_callback(az_log_message_fn az_log_message_callback)
 {
-  // TODO: thread safety
-  _az_log_listener = listener;
+  _az_log_message_callback = az_log_message_callback;
 }
 
-void az_log_write(az_log_classification classification, az_span message)
+void az_log_write(az_log_classification classification, char const* message, int32_t message_length)
 {
-  // TODO: thread safety
-  if (_az_log_listener != NULL && az_log_should_write(classification))
+  if (_az_log_message_callback != NULL && az_log_should_write(classification))
   {
-    (*_az_log_listener)(classification, message);
+    _az_log_message_callback(classification, message, message_length);
```

Comment:
This function should take the span like it used to.
Then, in the IF, this is where you put the code to 0-terminate the span and then convert it to a char * in order to call _az_log_message_callback.

Reply:
Please see the new iteration, I could've overengineered it a bit for the sake of safety, but that's logging code, so I hope we can do a few extra "if" statements.
 
But the significant thing is that now AZ_SPAN_LITERAL_FROM_STR()'s span capacity (not length!) now includes the zero terminator. Which is not untrue.

I am doing my nest effort to prevent AV when logging function gets invoked with the string literal that resides in the read-only memory. It is now guaranteed safe to call it with a literal defined with AZ_SPAN_LITERAL_FROM_STR(). Realistically, I think it will cover the reasonable cases - for dynamical content, you have modifiable buffer, for static messages you have string literals.

Reply:
I've achieved what I am describing above in a different way.